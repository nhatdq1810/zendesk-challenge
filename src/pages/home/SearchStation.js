import React from 'react';
import PropTypes from 'prop-types';
import { AutoComplete } from 'antd';
import styles from './SearchStation.module.scss';

export default function SearchStation({
  dataSrcStations, searchOriginStation, setOriginStation,
  searchDestStation, setDestStation,
}) {
  return (
    <div className={styles.wrapper}>
      <AutoComplete
        autoFocus
        className={styles.input}
        dataSource={dataSrcStations}
        onSearch={searchOriginStation}
        onSelect={setOriginStation}
        placeholder="Choose origin station"
      />
      <AutoComplete
        className={styles.input}
        dataSource={dataSrcStations}
        onSearch={searchDestStation}
        onSelect={setDestStation}
        placeholder="Choose destination station"
      />
    </div>
  );
}

SearchStation.propTypes = {
  dataSrcStations: PropTypes.arrayOf(PropTypes.string).isRequired,
  searchOriginStation: PropTypes.func.isRequired,
  setOriginStation: PropTypes.func.isRequired,
  searchDestStation: PropTypes.func.isRequired,
  setDestStation: PropTypes.func.isRequired,
};
