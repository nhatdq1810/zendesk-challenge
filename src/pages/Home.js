import React, { Component } from 'react';
import { Empty } from 'antd';
import styles from './Home.module.scss';
import stations from '../utils/stations.json';
import {
  getAdjStations, searchStations, getRoutePoints, getOrderedRoutes,
} from './Home.helper';
import SearchStation from './home/SearchStation';
import RoutesView from './home/RoutesView';

class Home extends Component {
  constructor(props) {
    super(props);
    this.state = {
      dataSrcStations: [],
      originStation: undefined,
      destStation: undefined,
      orderedRoutes: [],
      orderedRoutesByLines: [],
    };
    this.stationNames = Object.keys(stations);
  }

  updateStateAfterSearch = (value, key) => {
    this.setState({
      dataSrcStations: searchStations(this.stationNames, value),
      [key]: value,
    });
  }

  searchOriginStation = (value) => {
    this.updateStateAfterSearch(value, 'originStation');
  }

  searchDestStation = (value) => {
    this.updateStateAfterSearch(value, 'destStation');
  }

  setOriginStation = (value) => {
    this.setState(({ destStation }) => {
      let newState = { originStation: value };

      if (value && destStation && stations[value] && stations[destStation]) {
        newState = { ...newState, ...this.getRoutes(value, destStation) };
      }

      return newState;
    });
  }

  setDestStation = (value) => {
    this.setState(({ originStation }) => {
      let newState = { destStation: value };

      if (value && originStation && stations[value] && stations[originStation]) {
        newState = { ...newState, ...this.getRoutes(originStation, value) };
      }

      return newState;
    });
  }

  getRoutesUtil = (routes, routesByLines, originLines, destLine) => {
    if (originLines.some((originLine) => {
      if (originLine === destLine) {
        routes.push([...this.currentRoute]);
        routesByLines.push([...this.currentRouteByLines, destLine]);
        return true;
      }
      return false;
    })) { return; }

    originLines.forEach((originLine) => {
      const { adjStations, visitedStations } = getAdjStations(stations, originLine, this.visitedStations);
      this.visitedStations = [...visitedStations];
      this.currentRouteByLines.push(originLine);

      if (adjStations.length > 0) {
        adjStations.forEach((s) => {
          this.currentRoute.push({ name: s, ...stations[s] });
          const adjOriginLines = Object.keys(stations[s]);
          this.getRoutesUtil(routes, routesByLines, adjOriginLines, destLine);
          this.currentRoute.pop();
        });
      }

      this.currentRouteByLines.pop();
    });
  }

  getRoutes = (originStationName, destStationName) => {
    const originLines = Object.keys(stations[originStationName]);
    const routes = [];
    const routesByLines = [];
    this.visitedStations = [originStationName];

    Object.keys(stations[destStationName]).forEach((destLine) => {
      this.currentRoute = [];
      this.currentRouteByLines = [];
      this.getRoutesUtil(routes, routesByLines, originLines, destLine);
    });

    return getOrderedRoutes(
      routes,
      routesByLines,
      getRoutePoints(routes, routesByLines, stations[originStationName], stations[destStationName]),
    );
  }

  render() {
    const {
      dataSrcStations,
      orderedRoutes, orderedRoutesByLines,
      originStation, destStation,
    } = this.state;
    return (
      <div className={styles.wrapper}>
        <SearchStation
          dataSrcStations={dataSrcStations}
          searchOriginStation={this.searchOriginStation}
          setOriginStation={this.setOriginStation}
          searchDestStation={this.searchDestStation}
          setDestStation={this.setDestStation}
        />
        <div className={styles.routes}>
          {orderedRoutes.length === 0
            ? <Empty description="Please choose origin station and destination station" />
            : (
              <RoutesView
                orderedRoutes={orderedRoutes}
                orderedRoutesByLines={orderedRoutesByLines}
                originStation={originStation}
                destStation={destStation}
              />
            )
          }
        </div>
      </div>
    );
  }
}

export default Home;
