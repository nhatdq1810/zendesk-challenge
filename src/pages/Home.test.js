import React from 'react';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';
import Home from './Home';
import SearchStation from './home/SearchStation';

const createNodeMock = (el) => {
  if (el.type === 'input') {
    return { focus: jest.fn() };
  }
  return null;
};

jest.mock('../utils/stations.json', () => ({
  A: { NS: 1, CC: 2 },
  A1: { NS: 3 },
  B: { NS: 2, DT: 5 },
  C: { DT: 10, SE: [2, 15] },
  D: { SE: 3, EW: 4 },
  E: { EW: 3, CC: 5 },
}));

describe('Home', () => {
  let wrapper;

  beforeEach(() => {
    wrapper = shallow(<Home />);
  });

  it('should render correctly when orderedRoutes is empty', () => {
    const snapshot = renderer.create(<Home />, { createNodeMock }).toJSON();
    expect(snapshot).toMatchSnapshot();
  });

  it('should pass prop searchOriginStation correctly to SearchStation', () => {
    const searchStationComp = wrapper.find(SearchStation);
    const expectedOriginStation = 'expectedOriginStation';

    searchStationComp.prop('searchOriginStation')(expectedOriginStation);

    expect(wrapper.state('originStation')).toBe(expectedOriginStation);
  });

  it('should pass prop searchDestStation correctly to SearchStation', () => {
    const searchStationComp = wrapper.find(SearchStation);
    const expectedDestStation = 'expectedDestStation';

    searchStationComp.prop('searchDestStation')(expectedDestStation);

    expect(wrapper.state('destStation')).toBe(expectedDestStation);
  });

  it('should pass prop setOriginStation correctly to SearchStation', () => {
    const searchStationComp = wrapper.find(SearchStation);
    const expectedOriginStation = 'A1';

    searchStationComp.prop('setOriginStation')(expectedOriginStation);

    expect(wrapper.state('originStation')).toBe(expectedOriginStation);
    expect(wrapper.state('orderedRoutes')).toEqual([]);

    wrapper.setState({ destStation: 'B' });
    searchStationComp.prop('setOriginStation')(expectedOriginStation);

    expect(wrapper.state('originStation')).toBe(expectedOriginStation);
  });

  it('should pass prop setDestStation correctly to SearchStation', () => {
    const searchStationComp = wrapper.find(SearchStation);
    const expectedDestStation = 'A1';

    searchStationComp.prop('setDestStation')(expectedDestStation);

    expect(wrapper.state('destStation')).toBe(expectedDestStation);
    expect(wrapper.state('orderedRoutes')).toEqual([]);

    wrapper.setState({ originStation: 'B' });
    searchStationComp.prop('setDestStation')(expectedDestStation);

    expect(wrapper.state('destStation')).toBe(expectedDestStation);
  });
});

describe('getRoutes', () => {
  let wrapper;

  beforeEach(() => {
    wrapper = shallow(<Home />);
  });

  it('should return correct routes when origin and destination have same line', () => {
    const originStation = 'A';
    const destStation = 'A1';

    const routes = wrapper.instance().getRoutes(originStation, destStation);

    expect(routes).toEqual({ orderedRoutes: [[]], orderedRoutesByLines: [['NS']] });
  });

  it('should return correct routes when origin and destination do not have same line', () => {
    const originStation = 'A';
    const destStation = 'E';

    const routes = wrapper.instance().getRoutes(originStation, destStation);

    expect(routes).toEqual({
      orderedRoutes: [
        [{ EW: 3, CC: 5, name: 'E' }],
        [],
        [{ DT: 5, NS: 2, name: 'B' }, { DT: 10, SE: [2, 15], name: 'C' }, { EW: 4, SE: 3, name: 'D' }]],
      orderedRoutesByLines: [
        ['CC', 'EW'],
        ['CC'],
        ['NS', 'DT', 'SE', 'EW'],
      ],
    });
  });
});
