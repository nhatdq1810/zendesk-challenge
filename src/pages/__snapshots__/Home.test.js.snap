// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home should render correctly when orderedRoutes is empty 1`] = `
<div
  className="wrapper"
>
  <div
    className="wrapper"
  >
    <div
      className="input ant-select-show-search ant-select-auto-complete ant-select ant-select-combobox ant-select-enabled"
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onMouseDown={[Function]}
      onMouseOut={[Function]}
      onMouseUp={[Function]}
      onTouchStart={[Function]}
    >
      <div
        aria-autocomplete="list"
        aria-controls="test-uuid"
        aria-expanded={false}
        aria-haspopup="true"
        className="ant-select-selection
            ant-select-selection--single"
        role="combobox"
      >
        <div
          className="ant-select-selection__rendered"
        >
          <div
            className="ant-select-selection__placeholder"
            onClick={[Function]}
            onMouseDown={[Function]}
            style={
              Object {
                "WebkitUserSelect": "none",
                "display": "block",
                "userSelect": "none",
              }
            }
            unselectable="on"
          >
            Choose origin station
          </div>
          <ul>
            <li
              className="ant-select-search ant-select-search--inline"
            >
              <div
                className="ant-select-search__field__wrap"
              >
                <input
                  className="ant-input ant-select-search__field"
                  disabled={false}
                  onChange={[Function]}
                  onKeyDown={[Function]}
                  type="text"
                  value=""
                />
                <span
                  className="ant-select-search__field__mirror"
                >
                  
                   
                </span>
              </div>
            </li>
          </ul>
        </div>
        <span
          className="ant-select-arrow"
          onClick={[Function]}
          style={
            Object {
              "WebkitUserSelect": "none",
              "userSelect": "none",
            }
          }
          unselectable="on"
        >
          <i
            className="anticon anticon-down ant-select-arrow-icon"
          >
            <svg
              aria-hidden="true"
              className=""
              data-icon="down"
              fill="currentColor"
              height="1em"
              viewBox="64 64 896 896"
              width="1em"
            >
              <path
                d="M884 256h-75c-5.1 0-9.9 2.5-12.9 6.6L512 654.2 227.9 262.6c-3-4.1-7.8-6.6-12.9-6.6h-75c-6.5 0-10.3 7.4-6.5 12.7l352.6 486.1c12.8 17.6 39 17.6 51.7 0l352.6-486.1c3.9-5.3.1-12.7-6.4-12.7z"
              />
            </svg>
          </i>
        </span>
      </div>
    </div>
    <div
      className="input ant-select-show-search ant-select-auto-complete ant-select ant-select-combobox ant-select-enabled"
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onMouseDown={[Function]}
      onMouseOut={[Function]}
      onMouseUp={[Function]}
      onTouchStart={[Function]}
    >
      <div
        aria-autocomplete="list"
        aria-controls="test-uuid"
        aria-expanded={false}
        aria-haspopup="true"
        className="ant-select-selection
            ant-select-selection--single"
        role="combobox"
      >
        <div
          className="ant-select-selection__rendered"
        >
          <div
            className="ant-select-selection__placeholder"
            onClick={[Function]}
            onMouseDown={[Function]}
            style={
              Object {
                "WebkitUserSelect": "none",
                "display": "block",
                "userSelect": "none",
              }
            }
            unselectable="on"
          >
            Choose destination station
          </div>
          <ul>
            <li
              className="ant-select-search ant-select-search--inline"
            >
              <div
                className="ant-select-search__field__wrap"
              >
                <input
                  className="ant-input ant-select-search__field"
                  disabled={false}
                  onChange={[Function]}
                  onKeyDown={[Function]}
                  type="text"
                  value=""
                />
                <span
                  className="ant-select-search__field__mirror"
                >
                  
                   
                </span>
              </div>
            </li>
          </ul>
        </div>
        <span
          className="ant-select-arrow"
          onClick={[Function]}
          style={
            Object {
              "WebkitUserSelect": "none",
              "userSelect": "none",
            }
          }
          unselectable="on"
        >
          <i
            className="anticon anticon-down ant-select-arrow-icon"
          >
            <svg
              aria-hidden="true"
              className=""
              data-icon="down"
              fill="currentColor"
              height="1em"
              viewBox="64 64 896 896"
              width="1em"
            >
              <path
                d="M884 256h-75c-5.1 0-9.9 2.5-12.9 6.6L512 654.2 227.9 262.6c-3-4.1-7.8-6.6-12.9-6.6h-75c-6.5 0-10.3 7.4-6.5 12.7l352.6 486.1c12.8 17.6 39 17.6 51.7 0l352.6-486.1c3.9-5.3.1-12.7-6.4-12.7z"
              />
            </svg>
          </i>
        </span>
      </div>
    </div>
  </div>
  <div
    className="routes"
  >
    <div
      className="ant-empty"
    >
      <div
        className="ant-empty-image"
      >
        <img
          alt="Please choose origin station and destination station"
          src="data:image/svg+xml;base64,PHN2ZyB3aWR0aD0iMTg0IiBoZWlnaHQ9IjE1MiIgeG1sbnM9Imh0dHA6Ly93d3cudzMub3JnLzIwMDAvc3ZnIj4KICA8ZyBmaWxsPSJub25lIiBmaWxsLXJ1bGU9ImV2ZW5vZGQiPgogICAgPGcgdHJhbnNmb3JtPSJ0cmFuc2xhdGUoMjQgMzEuNjcpIj4KICAgICAgPGVsbGlwc2UgZmlsbC1vcGFjaXR5PSIuOCIgZmlsbD0iI0Y1RjVGNyIgY3g9IjY3Ljc5NyIgY3k9IjEwNi44OSIgcng9IjY3Ljc5NyIgcnk9IjEyLjY2OCIvPgogICAgICA8cGF0aCBkPSJNMTIyLjAzNCA2OS42NzRMOTguMTA5IDQwLjIyOWMtMS4xNDgtMS4zODYtMi44MjYtMi4yMjUtNC41OTMtMi4yMjVoLTUxLjQ0Yy0xLjc2NiAwLTMuNDQ0LjgzOS00LjU5MiAyLjIyNUwxMy41NiA2OS42NzR2MTUuMzgzaDEwOC40NzVWNjkuNjc0eiIgZmlsbD0iI0FFQjhDMiIvPgogICAgICA8cGF0aCBkPSJNMTAxLjUzNyA4Ni4yMTRMODAuNjMgNjEuMTAyYy0xLjAwMS0xLjIwNy0yLjUwNy0xLjg2Ny00LjA0OC0xLjg2N0gzMS43MjRjLTEuNTQgMC0zLjA0Ny42Ni00LjA0OCAxLjg2N0w2Ljc2OSA4Ni4yMTR2MTMuNzkyaDk0Ljc2OFY4Ni4yMTR6IiBmaWxsPSJ1cmwoI2xpbmVhckdyYWRpZW50LTEpIiB0cmFuc2Zvcm09InRyYW5zbGF0ZSgxMy41NikiLz4KICAgICAgPHBhdGggZD0iTTMzLjgzIDBoNjcuOTMzYTQgNCAwIDAgMSA0IDR2OTMuMzQ0YTQgNCAwIDAgMS00IDRIMzMuODNhNCA0IDAgMCAxLTQtNFY0YTQgNCAwIDAgMSA0LTR6IiBmaWxsPSIjRjVGNUY3Ii8+CiAgICAgIDxwYXRoIGQ9Ik00Mi42NzggOS45NTNoNTAuMjM3YTIgMiAwIDAgMSAyIDJWMzYuOTFhMiAyIDAgMCAxLTIgMkg0Mi42NzhhMiAyIDAgMCAxLTItMlYxMS45NTNhMiAyIDAgMCAxIDItMnpNNDIuOTQgNDkuNzY3aDQ5LjcxM2EyLjI2MiAyLjI2MiAwIDEgMSAwIDQuNTI0SDQyLjk0YTIuMjYyIDIuMjYyIDAgMCAxIDAtNC41MjR6TTQyLjk0IDYxLjUzaDQ5LjcxM2EyLjI2MiAyLjI2MiAwIDEgMSAwIDQuNTI1SDQyLjk0YTIuMjYyIDIuMjYyIDAgMCAxIDAtNC41MjV6TTEyMS44MTMgMTA1LjAzMmMtLjc3NSAzLjA3MS0zLjQ5NyA1LjM2LTYuNzM1IDUuMzZIMjAuNTE1Yy0zLjIzOCAwLTUuOTYtMi4yOS02LjczNC01LjM2YTcuMzA5IDcuMzA5IDAgMCAxLS4yMjItMS43OVY2OS42NzVoMjYuMzE4YzIuOTA3IDAgNS4yNSAyLjQ0OCA1LjI1IDUuNDJ2LjA0YzAgMi45NzEgMi4zNyA1LjM3IDUuMjc3IDUuMzdoMzQuNzg1YzIuOTA3IDAgNS4yNzctMi40MjEgNS4yNzctNS4zOTNWNzUuMWMwLTIuOTcyIDIuMzQzLTUuNDI2IDUuMjUtNS40MjZoMjYuMzE4djMzLjU2OWMwIC42MTctLjA3NyAxLjIxNi0uMjIxIDEuNzg5eiIgZmlsbD0iI0RDRTBFNiIvPgogICAgPC9nPgogICAgPHBhdGggZD0iTTE0OS4xMjEgMzMuMjkybC02LjgzIDIuNjVhMSAxIDAgMCAxLTEuMzE3LTEuMjNsMS45MzctNi4yMDdjLTIuNTg5LTIuOTQ0LTQuMTA5LTYuNTM0LTQuMTA5LTEwLjQwOEMxMzguODAyIDguMTAyIDE0OC45MiAwIDE2MS40MDIgMCAxNzMuODgxIDAgMTg0IDguMTAyIDE4NCAxOC4wOTdjMCA5Ljk5NS0xMC4xMTggMTguMDk3LTIyLjU5OSAxOC4wOTctNC41MjggMC04Ljc0NC0xLjA2Ni0xMi4yOC0yLjkwMnoiIGZpbGw9IiNEQ0UwRTYiLz4KICAgIDxnIHRyYW5zZm9ybT0idHJhbnNsYXRlKDE0OS42NSAxNS4zODMpIiBmaWxsPSIjRkZGIj4KICAgICAgPGVsbGlwc2UgY3g9IjIwLjY1NCIgY3k9IjMuMTY3IiByeD0iMi44NDkiIHJ5PSIyLjgxNSIvPgogICAgICA8cGF0aCBkPSJNNS42OTggNS42M0gwTDIuODk4LjcwNHpNOS4yNTkuNzA0aDQuOTg1VjUuNjNIOS4yNTl6Ii8+CiAgICA8L2c+CiAgPC9nPgo8L3N2Zz4K"
        />
      </div>
      <p
        className="ant-empty-description"
      >
        Please choose origin station and destination station
      </p>
    </div>
  </div>
</div>
`;
